File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if(!root || !subRoot) return false ;
        if(isSameTree(root,subRoot)) return true ;
        return this.isSubtree(root.left,subRoot) || this.isSubtree(root.right,subRoot);
    }
}
function isSameTree( p, q) {
        if(p==null || q==null) return p == q ;
        return p.val == q.val && isSameTree(p.left,q.left) && isSameTree(p.right,q.right);
    }

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int max[] = new int[]{0};
        solve(root,max);
        return max[0] -1 ;
    }
    private int solve(TreeNode  node , int max[]){
        if(node == null) return 0 ;
        int left = solve(node.left,max);
        int right = solve(node.right,max);
        max[0] = Math.max(max[0],left+right+1);
        return 1 + Math.max(left,right);
    }
}
